base64ct/src/base64.rs

```diff
@@ -0,0 +1,229 @@
+pub fn encoded_len(bytes: &[u8]) -> usize {
+    (((bytes.len() * 4) / 3) + 3) & !3
+}
+
+/// Encode the input byte slice as a Base64-encoded [`String`].
+#[cfg(feature = "alloc")]
+#[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+pub fn encode_string(input: &[u8]) -> String {
+    let elen = encoded_len(input);
+    let mut dst = vec![0u8; elen];
+    let res = encode(input, &mut dst);
+    debug_assert_eq!(elen, res.unwrap());
+    debug_assert!(str::from_utf8(&dst).is_ok());
+    // SAFETY: `dst` is fully written and contains only valid one-byte UTF-8 chars
+    unsafe { String::from_utf8_unchecked(dst) }
+}
+
+/// Decode
+pub fn decode(src: impl AsRef<[u8]>, dst: &mut [u8]) -> Result<usize, InvalidEncodingError> {
+    let src = src.as_ref();
+
+    // TODO(tarcieri): constant-time whitespace tolerance
+    if !src.is_empty() && char::from(src[src.len() - 1]).is_whitespace() {
+        return Err(InvalidEncodingError);
+    }
+
+    debug_assert!(decoded_len(src) <= dst.len());
+
+    let mut src_offset: usize = 0;
+    let mut dst_offset: usize = 0;
+    let mut src_length: usize = src.len();
+    let mut err: isize = 0;
+
+    while src_length > 4 {
+        err |= decode_3bytes(
+            &src[src_offset..(src_offset + 4)],
+            &mut dst[dst_offset..(dst_offset + 3)],
+        );
+        src_offset += 4;
+        dst_offset += 3;
+        src_length -= 4;
+    }
+
+    if src_length > 0 {
+        let mut i = 0;
+        let mut tmp_out = [0u8; 3];
+        let mut tmp_in = [b'A'; 4];
+
+        while i < src_length && src[src_offset + i] != b'=' {
+            tmp_in[i] = src[src_offset + i];
+            i += 1;
+        }
+
+        if i < 2 {
+            err = 1;
+        }
+
+        src_length = i - 1;
+        err |= decode_3bytes(&tmp_in, &mut tmp_out);
+        // tmp_in.zeroize();
+
+        dst[dst_offset..(dst_offset + src_length)].copy_from_slice(&tmp_out[..src_length]);
+        // tmp_out.zeroize();
+
+        dst_offset += i - 1;
+    }
+
+    if err == 0 {
+        Ok(dst_offset)
+    } else {
+        Err(InvalidEncodingError)
+    }
+}
+
+/// Decode Base64 to a vec
+#[cfg(feature = "alloc")]
+#[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+pub fn decode_vec(input: &str) -> Result<Vec<u8>, InvalidEncodingError> {
+    let dlen = decoded_len(input.as_bytes());
+    let mut output = vec![0u8; dlen];
+    let res = decode(input, &mut output)?;
+    debug_assert_eq!(dlen, res);
+    Ok(output)
+}
+
+/// Decoded len
+pub fn decoded_len(bytes: &[u8]) -> usize {
+    if bytes.is_empty() {
+        return 0;
+    }
+
+    let mut i = bytes.len() - 1;
+    let mut pad_count: usize = 0;
+
+    while i > 0 && bytes[i] == b'=' {
+        pad_count += 1;
+        i -= 1;
+    }
+
+    ((bytes.len() - pad_count) * 3) / 4
+}
+
+// B64 character set:
+// [A-Z]      [a-z]      [0-9]      +     /
+// 0x41-0x5a, 0x61-0x7a, 0x30-0x39, 0x2b, 0x2f
+
+#[inline(always)]
+pub(crate) fn encode_3bytes(src: &[u8], dst: &mut [u8]) {
```

Comment:
Maybe it's worth to move these functions into a separate `utils` module?

Reply:
I wound up extracting them into a `common` module. There's probably more code that can be DRYed out between them.